### PR TITLE
adds JSON.MSET implementation

### PR DESCRIPTION
**Pull request by roshkhatri**
adds JSON.MSET implementation

### CODE REVIEW

File: src/json/json.cc [hunk 1]
```diff
@@ -259,6 +259,59 @@ STATIC JsonUtilCode parseSetCmdArgs(ValkeyModuleString **argv, const int argc, S
...
+    size_t num_keys = (argc - 1) / 3;
+
+    MSetCmdArgs *args_list = (MSetCmdArgs *)malloc(num_keys * sizeof(MSetCmdArgs));
+    if (args_list == NULL) {
+        return JSONUTIL_MEMORY_ERROR;
+    }
+
+    memset(args_list, 0, num_keys * sizeof(MSetCmdArgs));
+
+    if (args_list == NULL) {
```
Comment (joehu21) [inline src/json/json.cc:288]: why checking if args_list is null again?

Comment (roshkhatri) [reply src/json/json.cc:288]: Oh. I forgot to remove it

File: src/json/json.cc [hunk 2]
```diff
@@ -259,6 +259,59 @@ STATIC JsonUtilCode parseSetCmdArgs(ValkeyModuleString **argv, const int argc, S
...
+} MSetCmdArgsList;
+
+STATIC JsonUtilCode parseMSetCmdArgs(ValkeyModuleString **argv, const int argc, MSetCmdArgsList *args) {
+    // we need 3 arguments for each key for MSET
+    if ((argc - 1) % 3) {
+        return JSONUTIL_WRONG_NUM_ARGS;
+    }
+    size_t num_keys = (argc - 1) / 3;
+
+    MSetCmdArgs *args_list = (MSetCmdArgs *)malloc(num_keys * sizeof(MSetCmdArgs));
```
Comment (joehu21) [inline src/json/json.cc:281]: Use ValkeyModule_Malloc instead of malloc.

File: src/json/json.cc [hunk 3]
```diff
@@ -259,6 +259,59 @@ STATIC JsonUtilCode parseSetCmdArgs(ValkeyModuleString **argv, const int argc, S
...
+
+    args->arg_list = args_list;
+    args->num_keys = num_keys;
+
+    return JSONUTIL_SUCCESS;
+}
+
+STATIC void freeMSetCmdArgs(MSetCmdArgsList *args) {
+    if (args != NULL && args->arg_list != NULL) {
+        free(args->arg_list);
```
Comment (joehu21) [inline src/json/json.cc:304]: Use ValkeyModule_Free

File: src/json/json.cc [hunk 4]
```diff
@@ -625,6 +678,116 @@ int Command_JsonSet(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
...
+            if (doc == nullptr) {
+                freeMSetCmdArgs(&args);
+                return ValkeyModule_ReplyWithError(ctx, ERRMSG_JSON_DOCUMENT_NOT_FOUND);
+            }
+        }
+    }
+
+    // Second phase: Apply changes
+    for (size_t i = 0; i < args.num_keys; i++) {
+        ValkeyModuleKey *key = static_cast<ValkeyModuleKey*>(ValkeyModule_OpenKey(ctx, args.arg_list[i].key,
```
Comment (joehu21) [inline src/json/json.cc:719]: Repeated operation. We are re-opening all keys in phase 2. Can we store the opened keys in phase 1 in an array and reuse it here.

File: src/json/json.cc [hunk 5]
```diff
@@ -625,6 +678,116 @@ int Command_JsonSet(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
...
+                                                                               VALKEYMODULE_READ | VALKEYMODULE_WRITE));
+        bool is_root_path = jsonutil_is_root_path(args.arg_list[i].path);
```
Comment (joehu21) [inline src/json/json.cc:721]: same here. we checked twice if each path is root.

File: src/json/json.cc [hunk 6]
```diff
@@ -625,6 +678,116 @@ int Command_JsonSet(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
...
+            jsonstats_update_stats_on_insert(doc, true, 0, doc_size, doc_size);
+        } else {
+            JDocument *doc = static_cast<JDocument*>(ValkeyModule_ModuleTypeGetValue(key));
+            size_t orig_doc_size = dom_get_doc_size(doc);
+            
+            rc = dom_set_value(ctx, doc, args.arg_list[i].path, args.arg_list[i].json, 
+                             args.arg_list[i].json_len, false, false);  // no create/update only flags
+            if (rc != JSONUTIL_SUCCESS) {
+                freeMSetCmdArgs(&args);
+                return ValkeyModule_ReplyWithError(ctx, jsonutil_code_to_message(rc));
```
Comment (joehu21) [inline src/json/json.cc:754]: again, the atomicity is broken.

File: src/json/json.cc [hunk 7]
```diff
@@ -625,6 +678,116 @@ int Command_JsonSet(ValkeyModuleCtx *ctx, ValkeyModuleString **argv, int argc) {
...
+            // end tracking memory
+            int64_t delta = jsonstats_end_track_mem(begin_val);
+            size_t new_doc_size = dom_get_doc_size(doc) + delta;
+            dom_set_doc_size(doc, new_doc_size);
+
+            // update stats
+            jsonstats_update_stats_on_update(doc, orig_doc_size, new_doc_size, args.arg_list[i].json_len);
+        }
+
+        ValkeyModule_NotifyKeyspaceEvent(ctx, VALKEYMODULE_NOTIFY_GENERIC, "json.mset", args.arg_list[i].key);
```
Comment (joehu21) [inline src/json/json.cc:766]: this line should be moved to the place after L786.

**Review by joehu21 — COMMENTED** [review]
Can you add a pytest?

File: src/json/json.cc (hunk 5 above)
Comment (roshkhatri) [reply src/json/json.cc:712]: these are two for loops to verify the keys and the path and the other loop is to set the values

**Comment by roshkhatri** [comment]
I think we also need to check to all the values using `prepareSetValues` and the path and the values before setting anyone of them

**Pull request closed without merging**